File: zio-8807/src/mpsc/custom/Padded64.scala
package mpsc
package custom

import java.util.concurrent.atomic.AtomicReference

private[custom] abstract class Padded64Write[A]
    extends AtomicReference(List.empty[A])
    with Serializable

private[custom] abstract class Padded64WritePad[A] extends Padded64Write[A] {
  protected val _0 = 0L;
  protected val _1 = 0L;
  protected val _2 = 0L;
  protected val _3 = 0L;
  protected val _4 = 0L;
  protected val _5 = 0L;
  protected val _6 = 0L;
  protected val _7 = 0L;
}

final class Padded64[A] extends Padded64WritePad[A] with Queue[A] {

  @transient private var read = List.empty[A]

  def add(head: A): Unit = {
    var tail = null.asInstanceOf[List[A]]
    var prev = get()
    do {
      tail = prev
      prev = compareAndExchange(tail, head :: tail)
    } while (prev ne tail)
  }

  def poll(): A = {
    var read = this.read
    if (read.isEmpty) {
      read = getAndSet(Nil)
      if (read eq Nil) return null.asInstanceOf[A]
      if (read.tail ne Nil) read = read.reverse
    }
    this.read = read.tail
    read.head
  }
}

File: zio-8807/src/mpsc/vyukov/Padded64.java
package mpsc.vyukov;

import java.io.Serializable;
import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;

import mpsc.Queue;

abstract class Padded64Write implements Serializable {
	protected transient volatile Padded64.Node write;
}

abstract class Padded64WritePad extends Padded64Write {
	protected long _0;
	protected long _1;
	protected long _2;
	protected long _3;
	protected long _4;
	protected long _5;
	protected long _6;
	protected long _7;
}

public final class Padded64<A> extends Padded64WritePad implements Queue<A> {

	private transient Node read;

	public Padded64() {
		read = write = new Node(null);
	}

	@Override
	public void add(A data) {
		Node next = new Node(data);
		Node prev = WRITE.getAndSet(this, next);
		Node.NEXT.lazySet(prev, next);
	}

	@SuppressWarnings("unchecked")
	@Override
	public A poll() {
		Node next = read.next;

		if (next == null)
			return null;

		read = next;
		Object data = next.data;
		next.data = null;
		return (A) (data);
	}

	static class Node {
		Object data;
		volatile Node next;

		Node(Object data) {
			this.data = data;
		}

		Node(Object data, Node next) {
			this.data = data;
			this.next = next;
		}

		static final AtomicReferenceFieldUpdater<Node, Node> NEXT = AtomicReferenceFieldUpdater.newUpdater(Node.class,
				Node.class, "next");
	}

	static final AtomicReferenceFieldUpdater<Padded64Write, Node> WRITE = AtomicReferenceFieldUpdater
			.newUpdater(Padded64Write.class, Node.class, "write");
}
